nncf/torch/quantization/weights_compression.py

```diff
@@ -0,0 +1,123 @@
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Dict, Optional
+
+import torch
+from torch import nn
+
+from nncf.torch.layers import NNCF_WRAPPED_USER_MODULES_DICT
+from nncf.torch.layers import NNCFEmbedding
+from nncf.torch.layers import NNCFLinear
+from nncf.torch.quantization.quantize_functions import get_scale_zp_from_input_low_input_high
+
+
+class WeightsDecompressor(nn.Module):
+    """Class decompression of compressed weights in forward pass
+
+    Atributes:
+        zero_point: zero point in quantization scheme
+        scale: scale in quantizatin scheme
+    """
+
+    def __init__(self, zero_point, scale):
+        super().__init__()
+        self.zero_point = zero_point
+        self.scale = scale
+
+    def forward(self, layer, op_arg):
+        w = layer.weight.type(dtype=self.scale.dtype)
+        layer.weight = (w - self.zero_point) * self.scale
+
+
+class WeightsFQ(nn.Module):
+    """Class provided fake quantize operation in forward pass
+
+    Atributes:
+        zero_point: zero point in quantization scheme
+        scale: scale in quantizatin scheme
+        axis: channel for quantization
+    """
+
+    def __init__(self, zero_point, scale, axis=0):
+        super().__init__()
+        self.zero_point = zero_point
+        self.scale = scale
+        self.axis = axis
+
+    def forward(self, layer, op_arg):
+        layer.weight = torch.fake_quantize_per_channel_affine(
+            layer.weight, self.scale, self.zero_point, self.axis, 0, 255
+        )
+
+
+def _insert_pre_compression_operations(
+    module: nn.Module, allowed_types: Dict, compress_weights=False, lewel_high=255
+) -> Optional[nn.Module]:
+    """
+    Insets weights compression with dequantization or quantization pre operation for Linear and Embedding layers.
+
+    :param module: The module to insert the weights compression.
+    :param compress_weights: Enables real compression of weights in Linear and Embedding layers.
+        If False inserts pytorch torch.fake_quantize_per_channel_affine(),
+        else compress weights to int8 and inserts custom dequantization.
+    :param allowed_types: list of allowed types for weights compression
+    :param lewel_high: highest  possible value of compressed weights (lower is 0 in assymetric quantization)
+    :return: The module with inserted operations. The module is not trainable if compress_weights is True.
+    """
+    for _, layer in module.named_children():
+        if not type(layer) in allowed_types:
+            _insert_pre_compression_operations(layer, allowed_types, compress_weights, lewel_high)
+            continue
+        q_dim = layer.target_weight_dim_for_compression
```

Comment:
I'm not sure that these variables are clearly explainable. Can we rename them?

Reply:
Done